### PR TITLE
[RW-8054][risk=low] Fix ShareWorkspace for CT published workspaces

### PR DESCRIPTION
**Pull request by calbach**
- The fix: when running ShareWorkspace - we should always avoid removing the **RT** auth domain group, not necessarily the group associated with the CDR version. This is because PublishWorkspace always sets the RT group.
- Refactor shareWorkspace for clarity, I've filed https://precisionmedicineinitiative.atlassian.net/browse/RW-8065 for a longer-term fix here
- Performance improvement: don't do a group lookup, just use value from the config
- Some documentation fixes along the share code path

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/firecloud/FirecloudTransforms.java [hunk 1]
```diff
@@ -21,4 +23,30 @@ private FirecloudTransforms() {}
     Gson gson = new Gson();
     return gson.fromJson(gson.toJson(aclResp.getAcl(), accessEntryType), accessEntryType);
   }
+
+  public static FirecloudWorkspaceACLUpdate buildAclUpdate(
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/firecloud/FirecloudTransforms.java:27]: share and compute match the previous version

File: api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceAuthService.java [hunk 2]
```diff
@@ -99,101 +108,63 @@ public DbWorkspace getWorkspaceEnforceAccessLevelAndSetCdrVersion(
...
   public Map<String, FirecloudWorkspaceAccessEntry> getFirecloudWorkspaceAcls(
       String workspaceNamespace, String firecloudName) {
-    FirecloudWorkspaceACL aclResp =
-        fireCloudService.getWorkspaceAclAsService(workspaceNamespace, firecloudName);
-
-    // Swagger Java codegen does not handle the WorkspaceACL model correctly; it returns a GSON map
-    // instead. Run this through a typed Gson conversion process to parse into the desired type.
-    Type accessEntryType = new TypeToken<Map<String, FirecloudWorkspaceAccessEntry>>() {}.getType();
-    Gson gson = new Gson();
-    return gson.fromJson(gson.toJson(aclResp.getAcl(), accessEntryType), accessEntryType);
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceAuthService.java:137]: how did you resolve this nonsense?

File: api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceAuthService.java [hunk 3]
```diff
@@ -99,101 +108,63 @@ public DbWorkspace getWorkspaceEnforceAccessLevelAndSetCdrVersion(
...
+    // Note: keep in sync with WorkspaceAdminService.setPublished().
+    String publishedGroup = accessTierService.getRegisteredTierOrThrow().getAuthDomainGroupEmail();
+
     // Iterate through existing roles, update/remove them
     List<FirecloudWorkspaceACLUpdate> updateACLRequestList = new ArrayList<>();
     Map<String, WorkspaceAccessLevel> toAdd = new HashMap<>(updatedAclsMap);
     for (Map.Entry<String, FirecloudWorkspaceAccessEntry> entry : aclsMap.entrySet()) {
-      String currentUserEmail = entry.getKey();
-      WorkspaceAccessLevel updatedAccess = toAdd.get(currentUserEmail);
+      String email = entry.getKey();
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceAuthService.java:138]: thanks

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java [hunk 4]
```diff
@@ -2512,6 +2508,63 @@ public void testShareWorkspaceRemoveBillingProjectUser() {
...
+
+    DbUser writerUser = createAndSaveUser("writerfriend@gmail.com", 124L);
+
+    when(fireCloudService.getWorkspaceAclAsService(anyString(), anyString()))
+        .thenReturn(
+            createWorkspaceACL(
+                new JSONObject()
+                    .put(
+                        // Specifically, the REGISTERED tier is used for publishing.
+                        accessTier.getAuthDomainGroupEmail(),
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java:2524]: could also rename this to `registeredTier`

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java [hunk 5]
```diff
@@ -2512,6 +2508,63 @@ public void testShareWorkspaceRemoveBillingProjectUser() {
...
+        new ShareWorkspaceRequest()
+            .workspaceEtag(workspace.getEtag())
+            // Removing writer.
+            .addItemsItem(
+                new UserRole().email(LOGGED_IN_USER_EMAIL).role(WorkspaceAccessLevel.OWNER));
+
+    stubFcUpdateWorkspaceACL();
+    workspacesController.shareWorkspace(
+        workspace.getNamespace(), workspace.getName(), shareWorkspaceRequest);
+    verify(fireCloudService)
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java:2556]: a comment like "does not remove RT auth domain" would be helpful context here

**Review by jmthibault79 — APPROVED** [review]

**Comment by jmthibault79** [comment]
needs a lint btw

File: api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceAuthService.java (hunk 2 above)
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceAuthService.java:137]: It's still gross, I just centralized it - this was a duplication of a utility method we already had.

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java (hunk 4 above)
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java]: Good point, done

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java (hunk 5 above)
Comment (calbach) [reply api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java:2558]: Added

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 6]
```diff
@@ -684,8 +684,8 @@ private FirecloudWorkspaceId getFcBillingProject(DbAccessTier accessTier, Worksp
     final ImmutableMap<String, WorkspaceAccessLevel> aclsByEmail = shareRolesMapBuilder.build();
 
     List<String> workflowUsers = new ArrayList<>();
-    // Revoke lifesciene permission before asking Firecloud to remove users. Because after unshare
-    // in Firecloud, we nolonger can get user petSA from SAM using their credential.
+    // Revoke lifescience permission before asking Firecloud to remove users. Because after unshare
+    // in Firecloud, we no longer can get user petSA from SAM using their credential.
```
[the remark this answers is not in the document]
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java]: done

**Pull request merged**